pkg/autoscaler/aggregation/bucketing2.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2019 The Knative Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aggregation
+
+import (
+	"math"
+	"sync"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+)
+
+// TimedFloat64Buckets2 keeps buckets that have been collected at a certain time.
+type TimedFloat64Buckets2 struct {
+	bucketsMutex sync.RWMutex
+	buckets      []float64
+	lastWrite    time.Time
+
+	granularity time.Duration
+	window      time.Duration
+}
+
+func (t *TimedFloat64Buckets2) String() string {
+	return spew.Sdump(t.buckets)
+}
+
+// NewTimedFloat64Buckets2 generates a new TimedFloat64Buckets2 with the given
+// granularity.
+func NewTimedFloat64Buckets2(window, granularity time.Duration) *TimedFloat64Buckets2 {
+	// Number of buckets is `window` divided by `granularity`, rounded up.
+	// e.g. 60s / 2s = 30.
+	nb := int(math.Ceil(float64(window) / float64(granularity)))
+	return &TimedFloat64Buckets2{
+		buckets:     make([]float64, nb),
+		granularity: granularity,
+		window:      window,
+	}
+}
+
+// timeToIndex converts time to an integer that can be used for modulo
+// operations to find the index in the bucket list.
+// bucketMutex needs to be held.
+func (t *TimedFloat64Buckets2) timeToIndex(tm time.Time) int {
```

Comment:
`timeToIndex` seems like a misleading name to me. I can't come up with something better though 😂 . Should we make this a pure function to drop the requirement of holding the mutex?

Reply:
yeah, I thought about, so `timeToIndex` is the best I came up with. I can inline it.

Reply:
Eh, it's used in a few places, so I'd keep it for now.